libia2/include/ia2_compartment_init.inc

```diff
@@ -60,3 +65,135 @@ void COMPARTMENT_IDENT(init_tls)(void) {
   };
   dl_iterate_phdr(protect_tls_pages, &args);
 }
+
+/* Magic symbol from the linker that marks the start of the ELF headers */
+extern Elf64_Ehdr __ehdr_start __attribute__((weak));
+
+struct FinalizerInfo {
+  uint64_t fini_offset;
+  uint64_t fini_array_offset;
+  uint64_t fini_array_size;
+};
+
+/* Filled in by runtime initialization with contents of  .fini_array and
+ * .fini */
+static volatile struct FinalizerInfo finalizers
+    __attribute__((__section__(".data.rel.ro.1"))) = {0};
+
+extern void COMPARTMENT_IDENT(__wrap_ia2_compartment_destructor)(void);
+
+__attribute__((no_sanitize("undefined"))) __attribute__((used)) void
+COMPARTMENT_IDENT(ia2_compartment_destructor)(void) {
+  static bool ran_destructors = false;
+  if (!&__ehdr_start || ran_destructors)
```

Comment:
Is `!&__ehdr_start` valid or a malformed ELF (i.e. should we assert that it's true)?